modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandler.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.transport.netty4;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+
+import java.nio.channels.ClosedChannelException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+
+/**
+ * Channel handler that queues up writes it receives and tries to only flush bytes as they can be written by the backing channel.
+ * This is helpful in reducing heap usage with handlers like {@link io.netty.handler.ssl.SslHandler} that might otherwise themselves
+ * buffer a large amount of data when the channel is not able to physically execute writes immediately.
+ */
+public final class Netty4WriteThrottlingHandler extends ChannelDuplexHandler {
```

Comment:
This is the same as the existing code except that I had to drop the thread context assertions that do not hold for HTTP messages. I'd rather drop the assertions though than duplicate this code or make it more complicated so that it can selectively handle this situation as it's already very complex.
Also, assuming things go according to plan a bigger refactoring of this code path both for transport and HTTP is imminent so this seems like the right level of effort to stabilize more deployments asap.

Reply:
One thing to note here is that this class is only needed with the TLS transport nowadays. The copy bytes channel does what it is intended for already in the non-TLS case.
I put it in the normal transport for HTTP now anyway and would move it to the security transports in a follow-up.